edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planner/HealthRestartPlanner.cs

```diff
@@ -103,14 +107,20 @@ public async Task<Plan> CreateShutdownPlanAsync(ModuleSet current)
                     .ToArray());
                 processedDesiredMatchingCurrentModules.UnionWith(desiredMatchingCurrentModules.Select(x => x.Key));
 
-                commands.AddRange(await this.ProcessDesiredAndCurrentSets(priorityBasedDesiredSet, priorityBasedCurrentSet, runtimeInfo, moduleIdentities));
+                (IEnumerable<ICommand> upfrontImagePullCommands, IEnumerable<ICommand> commands) = await this.ProcessDesiredAndCurrentSets(priorityBasedDesiredSet, priorityBasedCurrentSet, runtimeInfo, moduleIdentities);
+                upfrontImagePullPlan.AddRange(upfrontImagePullCommands);
+                plan.AddRange(commands);
             }
 
-            Events.PlanCreated(commands);
-            return new Plan(commands);
+            // If we need to premept the constructured plan with image pull
+            // commands, do so here.
+            plan = upfrontImagePullPlan.Concat(plan).ToList();
```

Comment:
Want to call out that image pull commands will preempt all other commands if feature flag enabled, including removing dead containers and updating edge agent container. Synced with Anca and we thought this was a good approach.